### PR TITLE
Todi details

### PR DESCRIPTION
**Pull request by StongeEtienne**
Added comments and details to TODI functions.

### CODE REVIEW

**Comment by frheault** [comment]
@StongeEtienne When we discussed we said that big complex function should have the long documentation (Dipy) and the short simple/explicit function (like normalize_vector) could keep the short documentation (since it was almost acceptable not to have any).

File: scilpy/tractanalysis/todi.py [hunk 1]
```diff
@@ -90,15 +142,36 @@ def mask_todi(self, mask):
...
+        """Smooth orientations on the sphere.
+
+        Smooth the orientations / distribution on the sphere.
+        Important for priors construction of BST.
+
+        Parameters
+        ----------
+        order : int, optional
+            Exponent blurring factor, based on the dot product
+            (default 2).
```
Comment (GuillaumeTh) [inline scilpy/tractanalysis/todi.py]: Why 2 now ?

File: scilpy/tractanalysis/todi.py [hunk 2]
```diff
@@ -135,25 +208,76 @@ def smooth_todi_spatial(self, sigma=0.5):
...
+        """Spherical Harmonics (SH) coefficients of the TODI map
+
+        Compute the SH representation of the TODI map,
+        converting SF to SH with a smoothing factor.
+
+        Parameters
+        ----------
+        sh_basis : {None, 'tournier07', 'descoteaux07'}
+            ``None`` for the default DIPY basis,
+            ``tournier07`` for the Tournier 2007 [2]_ basis, and
```
Comment (GuillaumeTh) [inline scilpy/tractanalysis/todi.py]: Please add [1] and [2] references

File: scilpy/tractanalysis/todi.py [hunk 3]
```diff
@@ -135,25 +208,76 @@ def smooth_todi_spatial(self, sigma=0.5):
...
+        """
+        return sf_to_sh(self.todi, self.sphere, sh_order, sh_basis, smooth)
 
     def reshape_to_3d(self, img_voxelly_masked):
+        """Reshape a complex ravelled image to 3D.
+
+        Unravel a given unravel mask (1D), image (1D), SH/SF (2D)
+        to its original 3D shape (with a 4D for SH/SF).
+
+
```
Comment (GuillaumeTh) [inline scilpy/tractanalysis/todi.py]: Remove an enter

File: scilpy/tractanalysis/todi.py [hunk 4]
```diff
@@ -163,12 +287,35 @@ def reshape_to_3d(self, img_voxelly_masked):
...
-        assert(peak_img.shape[-1] == 3)
+        """Compute distance to peak map.
+
+        Compute the distance of the TODI map to peaks at each position,
+        in radian or degree.
+
+        Parameters
+        ----------
+        peak_img : numpy.ndarray (4D)
+            Peaks image,  as written by most of Scilpy scripts.
```
Comment (GuillaumeTh) [inline scilpy/tractanalysis/todi.py]: two space after the comma. Please remove one.

File: scilpy/tractanalysis/todi_util.py [hunk 5]
```diff
@@ -1,11 +1,22 @@
...
 from scipy.spatial.ckdtree import cKDTree
 
 
 def streamlines_to_segments(streamlines):
+    """Split streamlines into its segments.
+
+    Parameters
+    ----------
+    streamlines : list of numpy.ndarray
+        Dimensions of the reference image.
```
Comment (GuillaumeTh) [inline scilpy/tractanalysis/todi_util.py]: Wrong documentation

File: scilpy/tractanalysis/todi_util.py [hunk 6]
```diff
@@ -17,6 +28,18 @@ def streamlines_to_segments(streamlines):
 
 
 def streamlines_to_endpoints(streamlines):
+    """Equivalent to streamlines resampling to 2 points (first and last).
+
+    Parameters
+    ----------
+    streamlines : list of numpy.ndarray
+        Dimensions of the reference image.
```
Comment (GuillaumeTh) [inline scilpy/tractanalysis/todi_util.py]: Wrong documentation

File: scilpy/tractanalysis/todi_util.py [hunk 7]
```diff
@@ -26,6 +49,22 @@ def streamlines_to_endpoints(streamlines):
 
 
 def streamlines_to_pts_dir_norm(streamlines):
+    """Evaluate each segment: mid position, direction, length.
+
+    Parameters
+    ----------
+    streamlines : list of numpy.ndarray
+        Dimensions of the reference image.
```
Comment (GuillaumeTh) [inline scilpy/tractanalysis/todi_util.py]: Wrong documentation

File: scilpy/tractanalysis/todi.py (hunk 1 above)
Comment (StongeEtienne) [reply scilpy/tractanalysis/todi.py]: it's the default used in the BST paper, its also the one used in TODI

**Comment by frheault** [comment]
Everything is working just like before. So in term of functionalities it is a pass

File: scilpy/tractanalysis/todi.py [hunk 8]
```diff
@@ -32,20 +52,32 @@ def set_todi(self, mask, todi):
         self.todi = todi
 
     def compute_todi(self, streamlines, length_weights=True):
+        """Compute the TODI map.
+
+        At each voxel an histogram distribution of
+        the local streamlines orientations (TODI) is computed,
+        with a Spherical Function (SF) representation of directions.
```
[the remark this answers is not in the document]
Comment (GuillaumeTh) [reply scilpy/tractanalysis/todi.py]: Finally, you can just remove `with a Spherical Function (SF) representation of directions.`

**Comment by frheault** [comment]
Just realized that the output is float64, should it be set to float32 to be consistent with most other scripts (So SH can be read by Fibernav for example)

**Comment by arnaudbore** [comment]
For me everything seems to be okay, @StongeEtienne fix the last comments including the thing about optional outputs by @frheault.

**Comment by arnaudbore** [comment]
> Everything is working just like before. So in term of functionalities it is a pass
> 
> Only quick note, something I didn't see last time.
> In the script `scil_compute_todi.py` at some point you create a list `output_file_list` to verify if files exist. You can actually put all output in the optional parameters of the `assert_outputs_exist` function.

@StongeEtienne can you take care of this. Once it's going to be fixed, we will have a last round of review before merging.

**Comment by GuillaumeTh** [comment]
@StongeEtienne Can you fix the conflicts please.

**Comment by StongeEtienne** [comment]
@arnaudbore 
However, if I don't put them in a list, I cannot verify that "parser.error('No output to be done')"

I could stll put them in the list however :

**Pull request merged**